### DISCUSSION TITLE
Extracting one band from a multiband BSQ file?

### DISCUSSION THREAD

**Discussion by jumpjack** [ImageMagick/ImageMagick#5186] [Digital Image Processing]
The [attached](https://github.com/ImageMagick/ImageMagick/files/8814141/2p291489372xylb0ltp2356l2m1.zip) zip contains two files: the .IMG is the image, the .LBL is the "label" describing its format.
The important part of the label is this one:

> /* IMAGE DATA ELEMENTS */
> 
> OBJECT                            = IMAGE
>   INTERCHANGE_FORMAT              = BINARY
>   LINES                           = 1024
>   LINE_SAMPLES                    = 1024
>   SAMPLE_TYPE                     = IEEE_REAL
>   SAMPLE_BITS                     = 32
>   BANDS                           = 3
>   BAND_STORAGE_TYPE               = **[BAND_SEQUENTIAL](https://desktop.arcgis.com/en/arcmap/10.5/manage-data/raster-and-images/bil-bip-and-bsq-raster-files.htm)**
>   CHECKSUM                        = 1.57131e+09
>   FIRST_LINE                      = 1
>   FIRST_LINE_SAMPLE               = 1
>   MAXIMUM                         = 4016.0
>   MEAN                            = 1498.52
>   MEDIAN                          = 1450.0
>   MINIMUM                         = 0.0
>   SAMPLE_BIT_MASK                 = 2#0000111111111111#
>   STANDARD_DEVIATION              = 296.218
>   INVALID_CONSTANT                = (0.0,0.0,0.0)
>   MISSING_CONSTANT                = (0.0,0.0,0.0)
> END_OBJECT      
> 

It's a "Band SeQuential" image (BSQ); its size is 1024x1024. The file contains 3 different "bands", which represent X, Y and Z coordinates of a 3d object.
A direct conversion to "something visible" results in this:

![image](https://user-images.githubusercontent.com/1620953/171407639-e0b4c842-e100-4f47-aad6-90f5d2f15d65.png)

Instead, what I need to get is a greyscale image where each pixel has a gray value representing the Z value of the model, which is known as "depth map".

By loading a depth map together with a companion image in [this page](https://depthmapviewer.ugocapeto.com/), you get a 3d model.

Unfortunately this [BSQ format](https://desktop.arcgis.com/en/arcmap/10.5/manage-data/raster-and-images/bil-bip-and-bsq-raster-files.htm)  is not very clear to me, so I am not able to write a VBA macro which extracts the right values and stores them into a plain .bmp file. Can Imagemagick be used for this task? I am on Windows 10.

BSQ structure:
![image](https://user-images.githubusercontent.com/1620953/171408606-c35226a2-9c77-4e0a-b898-942963953e89.png)

BSQ example (text format): https://desktop.arcgis.com/en/arcmap/10.5/manage-data/raster-and-images/bsq-format-example.htm

![image](https://user-images.githubusercontent.com/1620953/171408690-d492bf56-edfb-4132-9493-e789b84ff751.png)

**Please note that .IMG file is actually not just binary data, it also contains the text label itself, which should be removed before processing.**

[2p291489372xylb0ltp2356l2m1.zip](https://github.com/ImageMagick/ImageMagick/files/8814141/2p291489372xylb0ltp2356l2m1.zip)

**Comment by snibgo** [comment]
If the pixel values are not compressed, so we have 32 bits floating-point per value, and the start of each band is known, IM can be used to extract each band as a grayscale image. Then they can be saved as BMP or PNG or TIFF or whatever format you want.

In that metadata, I can't see an offset into the image file that says where each band starts.

The command to read one band and write it to a PNG file could be something like this:
```
magick -size 1024x1024+45056 -define quantum:format=floating-point -depth 32 gray:2p291489372xylb0ltp2356l2m1.img x.png
```
This guesses that a band starts at byte 45056 in the file.

The result, x.png, looks like this:
![x](https://user-images.githubusercontent.com/33812368/171437738-2991ed63-4410-49a0-9405-8db8d8ba71db.png)

This is clearly wrong, but not entirely wrong.

**Comment by snibgo** [comment]
Here is a better guess:
```
magick -size 1024x1024+45056 -define quantum:format=floating-point -depth 32 -endian MSB gray:2p291489372xylb0ltp2356l2m1.img -auto-level -combine x.png
```
This is similar to your expected result.
![x](https://user-images.githubusercontent.com/33812368/171446218-7981551e-21b4-4367-b8ec-e51d8e46fea7.png)
It doesn't look much like Mars to me, but what do I know?

**Reply by jumpjack** [reply]
why do you get 9 images as output of a 3-bands image?
And how does "combine" option combine them? (anyway, why combining them after separating them?)

Second image is apparently the same of the preview which I posted... which is wrong: a single-band image is grayscale by definition.

It does not look like mars because visible image data are stored into a different file, this one only contains 3d coordinates of a point cloud.

**Comment by snibgo** [comment]
My first image, with 9, is wrong. Your data needs "-endian MSB ".

"-combine" combines the three grayscale images into one colour image. Remove that to get 3 grayscale images.

"-auto-level" re-scales the data to a nominal range of 0 to QuantumRange, so we can see the image. If we remove that, and write verbose info...
```
magick -size 1024x1024+45056 -define quantum:format=floating-point -depth 32 -endian MSB gray:2p291489372xylb0ltp2356l2m1.img -verbose info:
```
... we can see the data in the first image is from 0 to 121 times QuantumRange, the second image is from -81 times QuantumRange to zero, and the third image is from zero to 8.7 times QuantumRange.

**Comment by jumpjack** [comment]
Thanks, I am doing some tests but with no success as of now.
Anyway, if you want to experiment, this is the source of images and data:

https://pds-imaging.jpl.nasa.gov/search/?fq=ATLAS_SPACECRAFT_NAME%3Aspirit&fq=-ATLAS_THUMBNAIL_URL%3Abrwsnotavail.jpg&fq=ATLAS_INSTRUMENT_NAME%3Anavcam&fq=PLANET_DAY_NUMBER%3A%5B1500%20TO%201850%5D&q=*2n289977682*

The images with "XYL" in the name are the one I am investigating, and you can see the "visual" images they are associated too.
About the other images... I don't know exactly, I am studying them.

**Comment by snibgo** [comment]
That NASA web page doesn't work for me. It says "Results: 24" but "displaying 0 to 0 of 0".

I have shown how to extract data from the file you supplied, 2p291489372xylb0ltp2356l2m1.img. The second image in that file has entirely negative values. In your OP, you seem to want to save daa in a BMP file. But BMP cannot record negative values. You need to do some processing to make the values positive.

**Comment by jumpjack** [comment]
Fixed link above.

[New link.](https://pds-imaging.jpl.nasa.gov/search/?fq=ATLAS_SPACECRAFT_NAME%3Aspirit&fq=ATLAS_INSTRUMENT_NAME%3Anavcam&fq=-ATLAS_THUMBNAIL_URL%3Abrwsnotavail.jpg&q=*2n289977682*)

Example image:
![image](https://user-images.githubusercontent.com/1620953/171999200-4934e944-1fba-4aba-ba06-deede1ad3c86.png)
2n289977682e**ffb**0eop1965l0m1

Corresponding XYZ data:

![image](https://user-images.githubusercontent.com/1620953/171999279-289f232f-0b2e-4a86-8c41-6bc2283ccc32.png)
2n289977682**xyl**b0eop1965l0m1

Unfortunately I don't understand how to download the second image, I get a raw .IMG rather than the converted jpeg.

Anyway the values in .IMG are coordinates, I should just get the minimum value, "convert" it into zero and normalize all other values w.r.t. it. The result has to be interpreted as a greyscale.

**Comment by snibgo** [comment]
Okay, for that new image, consider the EFF first. The files are:
```
f:\web\im>dir 2n289977682effb0eop1965l0m1*

 Directory of f:\web\im

04/06/2022  14:37         2,138,112 2n289977682effb0eop1965l0m1.img
04/06/2022  14:37            23,048 2n289977682effb0eop1965l0m1.img.lbl
```

2n289977682effb0eop1965l0m1.img.lbl includes this:
```
/* IMAGE DATA ELEMENTS */

OBJECT                            = IMAGE
  INTERCHANGE_FORMAT              = BINARY
  LINES                           = 1024
  LINE_SAMPLES                    = 1024
  SAMPLE_TYPE                     = MSB_INTEGER
  SAMPLE_BITS                     = 16
  BANDS                           = 1
```
So there is one image, 1024x1024 pixels, each pixel a 16-bit (2-byte) MSB integer.

The image file has 2138112 bytes. Assuming the image is at the end of the file, where does it start?
```
2138112 - 1024*1024*2 = 40960.
```
So we can extract the image like this:
```
magick -size 1024x1024+40960 -endian MSB -depth 16 gray:2n289977682effb0eop1965l0m1.img -verbose info:

Image:
  Filename: gray:2n289977682effb0eop1965l0m1.img
  Format: GRAY (Raw gray samples)
  Class: DirectClass
  Geometry: 1024x1024+0+0
  Units: Undefined
  Colorspace: Gray
  Type: Grayscale
  Endianness: Undefined
  Depth: 16-bit
  Channel depth:
    Gray: 16-bit
  Channel statistics:
    Pixels: 1048576
    Gray:
      min: 250  (0.00381476)
      max: 3493 (0.0532998)
      mean: 1066.21 (0.0162693)
      median: 2712 (0.0413825)
      standard deviation: 244.931 (0.00373741)
      kurtosis: 3.68222
```
min is 0.00381476, and max is 0.0532998, on a scale from 0.0 to 1.0. So the image is very dark. To make it visible, we can use "-auto-level":
```
magick -size 1024x1024+40960 -endian MSB -depth 16 gray:2n289977682effb0eop1965l0m1.img -auto-level eff.png
```
eff.png is this, which looks reasonable:

![eff](https://user-images.githubusercontent.com/33812368/172007433-8ec05015-e319-43c5-bd9a-d7cbafa3493e.png)


Repeat that work for the XYL image:
```
04/06/2022  15:07        12,627,968 2n289977682xylb0eop1965l0m1.img
04/06/2022  15:07            23,122 2n289977682xylb0eop1965l0m1.img.lbl
```

```
/* IMAGE DATA ELEMENTS */

OBJECT                            = IMAGE
  INTERCHANGE_FORMAT              = BINARY
  LINES                           = 1024
  LINE_SAMPLES                    = 1024
  SAMPLE_TYPE                     = IEEE_REAL
  SAMPLE_BITS                     = 32
  BANDS                           = 3
```
So we have three images, 1024x1024, 32-bit (4-byte) floating-point. I assume these are also MSB.
```
12627968 - 3 * 1024*1024*4 = 45056
```
```
magick -size 1024x1024+45056 -define quantum:format=floating-point -depth 32 -endian MSB gray:2n289977682xylb0eop1965l0m1.img -verbose info:
```
This shows metadata for the three images. The first has min=0 and max=52.9824 on a scale of 0 to 1. So many pixels are "whiter than white". We can save the images in TIFF format (which retains the weird values), or use "-auto-scale" to compress the values into a visible range and same as PNG or BMP or whatever we want.
```
magick -size 1024x1024+45056 -define quantum:format=floating-point -depth 32 -endian MSB gray:2n289977682xylb0eop1965l0m1.img -auto-level xyl.png
```
This has made three files:
```
04/06/2022  15:32           926,200 xyl-0.png
04/06/2022  15:32           996,330 xyl-1.png
04/06/2022  15:32         1,093,600 xyl-2.png
```
For example, xyl-0.png is:


![xyl-0](https://user-images.githubusercontent.com/33812368/172007116-a5619502-d8b8-4a02-8588-dc5e6e989425.png)

**Comment by jumpjack** [comment]
I have 2 problems here:
- I cannot arbitrarily "autolevel" the values, these are scientific data, so I need to extract numbers as they are
- if I remove the autolevel option, I get  a black/white xyz-0 image, and very poor coloured -1 and -2

![image](https://user-images.githubusercontent.com/1620953/172042636-61a7e3c0-db70-43e6-8d6c-1ce848aefdcd.png)

In the meantime, I found[ a strange page](https://pds-imaging.jpl.nasa.gov/w10n/mer/spirit/mer2no_0xxx/data/sol1843/rdr/2n289977530xylb0eop1965l0m1.img/0/raster/data[0:510,0:100,0:3]?output=json) which converts online IMG file into text; it's quite slow in converting 1 million bytes into several millions characters, but after some experiments I ended up converting a part of the image, which gives an idea of how the 3d model should look: it's very elongated going away from the observer, and very low; ideally it would fit in a 100 x 10 x 1000 box.

You can view [my resulting 3d file](https://win98.altervista.org/space/exploration/3d/test.ply) into this [online point cloud viewer.](http://lidarview.com/) (select format "XYZ"; use CTRL+mouse to rotate around vertical axis, SHIFT+mouse for other axes).

![image](https://user-images.githubusercontent.com/1620953/172044217-418fb3b9-6714-41e1-9a10-45064db500fd.png)

**Comment by snibgo** [comment]
Okay, so if you don't want "-auto-level", you can save the data in a format that retains the "weird" values. That format could be TIFF, which is viewable in common software, or some text format. For text, I suggest the "ftxt:" format:
```
magick -size 1024x1024+45056 -define quantum:format=floating-point -depth 32 -endian MSB gray:2n289977682xylb0eop1965l0m1.img -evaluate Divide %[fx:QuantumRange] ftxt:xyl.txt
```
If the command is in a BAT script, double the percent sign. It takes 14 seconds on my laptop. It creates three files: xyl-0.txt, xyl-1.txt and xyl-2.txt.

There is one text line per pixel. The lines look like this:
```
92,175:25.1247
93,175:24.7696
94,175:24.8302
95,175:25.0853
96,175:25.0106
97,175:25.0061
```
The three numbers on each line are: pixel x-coordinate, pixel y-coordinate, floating-point value.

I don't know what transformation is needed to create a point cloud from the data.

**Comment by jumpjack** [comment]
To get a 3d file you have just to replace ":" by "," and import in Meshlab.... but I am not getting the expected results.
Then only way to have something good as of now is the [page](https://pds-imaging.jpl.nasa.gov/w10n/mer/spirit/mer2no_0xxx/data/sol1843/rdr/2n289977530xylb0eop1965l0m1.img/0/raster/data%5B0:510,0:100,0:3%5D?output=json) mentioned above.

**Comment by jumpjack** [comment]
I found another kind of file, with extension  ".ht": it should be a "VICAR height map":
[2mesh_1869_n_137_ffl_0_v1.ht](https://pds-imaging.jpl.nasa.gov/data/w10n/mer/spirit/mer2mw_0xxx/data/navcam/site0137/2mesh_1869_n_137_ffl_0_v1.ht)

It contains this label:

LBLSIZE=4000
FORMAT='REAL'
TYPE='IMAGE'
BUFSIZ=4000
DIM=3
RECSIZE=4000
ORG='BSQ'
NL=812
NS=1000
NB=2
N1=1000
N2=812
N3=2
NBB=0
INTFMT='LOW'
REALFMT='RIEEE'
X_AXIS_MINIMUM=-50.684558  
Y_AXIS_MINIMUM=-22.048969  
MAP_SCALE=0.050512

Format description: https://www-mipl.jpl.nasa.gov/PAG/public/vug/vug9.html

Useful data:

FORMAT= samples format (float, 32 bit)
NL = lines (812)
NS = samples per line (1000)
NB = bands (2)

I tried your command with several combinations of parameter, but I get mostly garbage, although "something" appears barely visible in the resulting images.

With size 812x1000+4000 I get garbage; the same changing 4000 to 3999 or 3998 or similar; changing it to 4352 (as from calculations) results in error. Also tried with LSB or MSB.

I can't find specific details on this file format; being it an height map, it should in theory have just 1 band for the height values, but it has 2 ones, so...? Maybe they should be combined in a 64 bit float?


You can find other files [here](https://pds-imaging.jpl.nasa.gov/search/?fq=PRODUCT_TYPE%3A%22terrain%20mesh%22&fq=ATLAS_MISSION_NAME%3A%22mars%20exploration%20rover%22&fq=PLANET_DAY_NUMBER%3A%5B1850%20TO%201870%5D&fq=ATLAS_SPACECRAFT_NAME%3Aspirit&fq=-ATLAS_THUMBNAIL_URL%3Abrwsnotavail.jpg&q=*%3A*) (the ones starting by "2mesh") and [here](https://pds-imaging.jpl.nasa.gov/data/mer/spirit/mer2mw_0xxx/data/navcam/site0137/).

**Comment by jumpjack** [comment]
add on:
Maybe I found the meaning of the bands: meaningful one is just first ones, the others are derived/calculated.

"(*.ht) - Height Map (DEM) derived from the XYZ RDR, stored as an image file. It's in VICAR
format but is not fully compliant to VICAR label specifications. It has 3 bands. Band 1
contains actual height data. Band 2 fills areas for which there is no actual data using
interpolation. Band 3 provides metric of how close a pixel value is to the actual data."

**Comment by jumpjack** [comment]
add on:
hex analysis shows that the data actually start at 4000 (0xFa0):

![image](https://user-images.githubusercontent.com/1620953/173183985-a79e3b7e-0b39-41e5-9ddc-c0ebbf323983.png)

**Comment by snibgo** [comment]
> With size 812x1000+4000 I get garbage ...

"-size" parameters are width, then height, then offset. See https://imagemagick.org/script/command-line-options.php#size . So you should have "-size 1000x812+4000".

Looking at the binary data, the values seem to be LSB, not MSB.

The file is too small to contain three bands at 1000x812 pixels, 4 bytes/pixel.

Without a proper definition of the format, we can only guess.

**Comment by jumpjack** [comment]
Thanks for the tip, now my result passed from...
![image](https://user-images.githubusercontent.com/1620953/173193579-1159adfa-0712-405b-a200-21c352fbff1b.png)


to...

![image](https://user-images.githubusercontent.com/1620953/173193591-fceba866-9c17-4715-b242-142eec2e3165.png)

Using:

>magick -size 1000x1299+4000 -define quantum:format=floating-point -depth 32 -endian MSB gray:2mesh_1854_np_136_ffl_374_v1.ht 2mesh_1854_np_136_ffl_374_v1.png

which makes more sense, as an height map should show the terrain as seen from above.
I found that the text that I quoted is from a more recent mission (MSL), which uses 3 bands, while the file is of an older mission (MER2), and the label says that the bands are actually just 2. Assuming that also in old mission the second band only  contained calculated data, I think I can focus on first band. 

But current result looks **almost** monochromatic, while I would expect a greyscale:

![image](https://user-images.githubusercontent.com/1620953/173194717-bec8385a-176c-48de-a68c-d66d25e19aaa.png)

Am I missing something of the data held in the 32 bit?

**Comment by jumpjack** [comment]
sorry, I posted example for another file....

**Comment by jumpjack** [comment]
Using LSB, this:
![image](https://user-images.githubusercontent.com/1620953/173195988-b388d5bb-dfeb-47ec-b5b5-0e61d7d5eda9.png)

becomes this:
![image](https://user-images.githubusercontent.com/1620953/173195976-24c37294-41bb-4905-94e6-72de7ff466ce.png)

... which seems nonsense, but with another image, this:
![image](https://user-images.githubusercontent.com/1620953/173196019-6920ae48-7fbf-495a-9f64-15edfd7db024.png)


becomes this:

![image](https://user-images.githubusercontent.com/1620953/173196103-2dcb1813-8aab-45cc-a6c2-900a55ce0e00.png)

which in second band (which, if I understand correctly, is an attempt to fill the holes) becomes this:

![image](https://user-images.githubusercontent.com/1620953/173196077-0896bb40-2862-49dc-acb8-90c600f34b0d.png)


So probably LSB is correct, but a lot of data appear to get lost in first image.

**Comment by jumpjack** [comment]
Apparently data don't get lost, they are just invisible to human eye... Using auto-level brings them back.

Final example commands:

>magick -size 1000x1299+4000 -define quantum:format=floating-point -depth 32 -endian LSB gray:2mesh_1854_np_136_ffl_374_v1.ht -auto-level 2mesh_1854_np_136_ffl_374_v1.png

>magick -size 1000x812+4000 -define quantum:format=floating-point -depth 32 -endian LSB gray:2mesh_1869_n_137_ffl_0_v1.ht -auto-level 2mesh_1869_n_137_ffl_0_v1.png

>magick -size 1000x999+4000 -define quantum:format=floating-point -depth 32 -endian LSB gray:2mesh_1869_n_137_ffl_0_v2.ht -auto-level 2mesh_1869_n_137_ffl_0_v2.png


I have two further questions:
1) given this visual aspect of data, it is possible that actually they are 24 bit, not 32 bit, being 0x7fffff a "limit value"; but replacing 32 by 24 in command line makes command fail. Suggestions to try extracting 24bit data rather than 32?
Or maybe these files are not BSQ but [BIL or BIP format](https://desktop.arcgis.com/en/arcmap/10.5/manage-data/raster-and-images/bil-bip-and-bsq-raster-files.htm), 16 bit length (hence ff.ff.ff.7f would wplit into ffff for first band  and 7fff for second one; in such cases, which would be the CLI?


![immagine](https://user-images.githubusercontent.com/1620953/173380899-893168fd-9c21-4b31-8431-8b32af5ae392.png)

2) Is IM able to read in someway the NS and NL data in the label by itself?





Files:
https://pdsimage2.wr.usgs.gov/archive/mer2mw/mer2mw_0xxx/data/navcam/site0136/2mesh_1854_np_136_ffl_374_v1.ht

Label:

> 
> LBLSIZE=4000
> BUFSIZ=4000
> RECSIZE=4000
> FORMAT='REAL'
> TYPE='IMAGE'
> ORG='BSQ'
> NBB=0
> DIM=3
> INTFMT='LOW'
> REALFMT='RIEEE'
> **NS=1000**
> N1=1000
> **NL=1299**
> N2=1299
> NB=2
> N3=2 X_AXIS_MINIMUM=-102.386051
> Y_AXIS_MINIMUM=-271.654668
> MAP_SCALE=0.221833



https://pdsimage2.wr.usgs.gov/archive/mer2mw/mer2mw_0xxx/data/navcam/site0137/2mesh_1869_n_137_ffl_0_v1.ht

Label:

> LBLSIZE=4000
> BUFSIZ=4000
> RECSIZE=4000
> FORMAT='REAL'
> TYPE='IMAGE'
> ORG='BSQ'
> NBB=0
> DIM=3
> INTFMT='LOW'
> REALFMT='RIEEE'
> **NS=1000**
> N1=1000
> **NL=812**
> N2=812
> NB=2
> N3=2 X_AXIS_MINIMUM=-50.684558
> Y_AXIS_MINIMUM=-22.048969
> MAP_SCALE=0.050512



https://pdsimage2.wr.usgs.gov/archive/mer2mw/mer2mw_0xxx/data/navcam/site0137/2mesh_1869_n_137_ffl_0_v2.ht


Label:

> LBLSIZE=4000
> BUFSIZ=4000
> RECSIZE=4000
> FORMAT='REAL'
> TYPE='IMAGE'
> ORG='BSQ'
> NBB=0
> DIM=3
> INTFMT='LOW'
> REALFMT='RIEEE'
> NS=1000
> N1=1000
> NL=999
> N2=999
> NB=2
> N3=2 X_AXIS_MINIMUM=-50.891429
> Y_AXIS_MINIMUM=-90.690873
> MAP_SCALE=0.109683

**Comment by snibgo** [comment]
> given this visual aspect of data, it is possible that actually they are 24 bit

No. The bytes seem to be in groups of four, not two or three.

> Is IM able to read in someway the NS and NL data in the label by itself?

Yes, but that won't help you much. These files are something like the "Vicar" format. IM thinks they are that format, and can read that metadata. However, IM assumes each file has one image, and there seems to be no metadata in 2mesh_1869_n_137_ffl_0_v1.ht that says there are two images. In addition, IM thinks the file is 8-bit integer.

You could use IM to get the width and height, but would still need to calculate the correct byte offset, which depends on the number of images and filesize.

Instead, I suggest extracting that metadata (NS and NL) using a shell script.

**Reply by jumpjack** [reply]
> > given this visual aspect of data, it is possible that actually they are 24 bit
> 
> No. The bytes seem to be in groups of four, not two or three.

Sorry, I wrote it incorrectly:
I thought that "ff ff 7f 7f" could mean " 'ff ff' for first image, '7f 7f' for second image". The parameter "ORG='BSQ'" says it's a BSQ file, but I suspected it could be a BIP file:

![image](https://user-images.githubusercontent.com/1620953/173409474-10f6cac4-72a3-4343-92a8-9ea90fab4b6f.png)

But actually the REALFMT parameter confirms each sample is 32 bit.
.
.
.
.
> 
> > Is IM able to read in someway the NS and NL data in the label by itself?
> 
> Yes, but that won't help you much. These files are something like the "Vicar" format. IM thinks they are that format, and can read that metadata. However, IM assumes each file has one image, and there seems to be no metadata in 2mesh_1869_n_137_ffl_0_v1.ht that says there are two images. In addition, IM thinks the file is 8-bit integer.

The number of images is given by the number of bands, which in VICAR label is given by line "NB=2".

> 
> You could use IM to get the width and height, but would still need to calculate the correct byte offset, which depends on the number of images and filesize.

The "official" way to calculate the byte offset is reading LBLSIZE parameter of VICAR label. Hence "magick -size 1000x1299+4000 ..." would be  "magick -size NSxNL+LBLSIZE ..."

Commented (by me) VICAR label as per [this document](https://www-mipl.jpl.nasa.gov/external/VICAR_file_fmt.pdf):

LBLSIZE=4000 - Total length of the label = image start offset
BUFSIZ=4000 - Obsolete, always = RECSIZE
RECSIZE=4000 - Bytes per line (=pixel per line   x   bytes per pixel  = NS   x  FORMAT)   (if FORMAT then  REAL = 4)
FORMAT='REAL' - Single Precision Flpat (= 32 bit)
TYPE='IMAGE'
ORG='BSQ' - Band SeQuential 
NBB=0 - record prefix length
DIM=3 - constant
**INTFMT**='LOW' - If used format is REAL, INTFMT says if it's LSB or MSB:  HIGH = big endian, **LOW = little endian**
REALFMT='RIEEE' - RIEEE: Reverse [IEEE format](https://it.wikipedia.org/wiki/IEEE_754) (32 bit). Just like IEEE, except the bytes are reversed, with the exponent last.
NS=1000 - The number of samples in the image  
N1=1000 - as NS in BSQ
NL=999 - The number of lines in the image
N2=999 - as NL in BSQ
NB=2 - Number of images or bands
N3=2  - as NB in BSQ
X_AXIS_MINIMUM=-50.891429
Y_AXIS_MINIMUM=-90.690873
MAP_SCALE=0.109683

Last 3 parameters are explained [here](https://pds-imaging.jpl.nasa.gov/data/nsyt/insight_cameras/document/pds4_attribute_definitions_sort_pds.html#cart:x_axis_maximum).

**Comment by snibgo** [comment]
Thanks for the link to the VICAR file format.

From the IM source coders\vicar.c, IM ignores "NB", "ORG", "FORMAT" and most other metadata.

**Comment by jumpjack** [comment]
Thanks,  I opened an issue: https://github.com/ImageMagick/ImageMagick/issues/5221
